alchemist/alchemist-euclidean-geometry/src/main/kotlin/it/unibo/alchemist/model/interfaces/geometry/euclidean2d/Line2D.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2010-2020, Danilo Pianini and contributors
+ * listed in the main project's alchemist/build.gradle.kts file.
+ *
+ * This file is part of Alchemist, and is distributed under the terms of the
+ * GNU General Public License, with a linking exception,
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+
+package it.unibo.alchemist.model.interfaces.geometry.euclidean2d
+
+import it.unibo.alchemist.model.interfaces.geometry.Vector2D
+
+/**
+ * Defines a straight line in a cartesian plane.
+ */
+interface Line2D<P : Vector2D<P>> {
+
+    /**
+     * Indicates if the line is aligned to the x-axis.
+     */
+    val isHorizontal: Boolean
+
+    /**
+     * Indicates if the line is aligned to the y-axis.
+     */
+    val isVertical: Boolean
+
+    /**
+     * The slope of the line, if [isVertical] this is undefined (= [Double.NaN]).
+     */
+    val slope: Double
+
+    /**
+     * The y-coordinate of the y-intercept (= the point where the line intersects the y-axis). If [isVertical] there's
+     * no y-intercept and this is [Double.NaN]. The slope-intercept representation (y = mx + b) uses this intercept.
+     */
+    val yIntercept: Double
+
+    /**
+     * The x-coordinate of the x-intercept (= the point where the line intersects the x-axis). If [isHorizontal]
+     * there's no x-intercept and this is [Double.NaN].
+     */
+    val xIntercept: Double
+
+    /**
+     * Checks if the [point] belongs to this line.
+     */
+    fun contains(point: P): Boolean
+
+    /**
+     * Finds the point belonging to the line with the given [x]-coordinate. Throws an [UnsupportedOperationException]
+     * if the line [isVertical].
+     */
+    fun findPoint(x: Double): P
+
+    /**
+     * Checks if two lines are parallel.
+     */
+    fun isParallelTo(other: Line2D<*>): Boolean
```

Comment:
Why projected?

Reply:
That function is used in the implementation of `equals(other)`, where it's handy to have it out-projected (as I can only check `other is Line2D<*>`).

Reply:
that's missing implementation concerns with interface definition. If the contract has no reason to accept projected types, then it should accept the correct type.
I believe equals can be implemented in several ways (e.g., same x and y intercepts), am I wrong?

Reply:
You're not, even though it's handy to implement it using `isParallelTo`. Anyway, I corrected it.